File: 0739-daily-temperatures/0739-daily-temperatures.java
class Solution {
    public int[] dailyTemperatures(int[] temp) {
        int n = temp.length;
        int[] res = new int[n];
        
        Deque<Integer>st = new ArrayDeque<>();
        st.push(n - 1);
        
        for(int i = n - 2; i >= 0; i--){
            
            while(!st.isEmpty() && temp[st.peek()] <= temp[i])
                st.pop();
            
            res[i] = (st.isEmpty() ? 0 : st.peek() - i);
            
            st.push(i);            
        }
        
        return res;
    }
}

File: 0739-daily-temperatures/0739-daily-temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temp) {
        int n = temp.size();
        vector<int>res(n);
        
        stack<int>st;
        
        for(int i = n - 1; i >= 0; i--){
            
            while(!st.empty() && temp[st.top()] <= temp[i])
                st.pop();
            
            res[i] = (st.empty() ? 0 : st.top() - i);
            
            st.push(i);            
        }
        
        return res;  
    }
};